File: Server/src/main/java/fr/nathan/mim/api/data/DataObject.java
package fr.nathan.mim.api.data;

import fr.nathan.mim.Main;

import java.util.HashMap;

public class DataObject extends DataElement {

    protected HashMap<String, DataElement> children = new HashMap<>();

    public DataObject()                               {}
    public DataObject(DataObject copy)                {copy.children.forEach(this::put);}

    public boolean has(String key)                    {return children.containsKey(key);}
    public void put(String key, DataElement element)  {children.put(key, element.clone());}
    public DataElement get(String key)                {return children.get(key);}
    public HashMap<String, DataElement> getChildren() {return children;}

    @Override
    public DataElement clone() {return new DataObject(this);}

    @Override
    public String toString() {
        return Main.currentParser.toString(this);
    }

}

File: Client/src/data/DataObject.cpp
//
// Created by Nathan David on 02/02/2022.
//

#include "data/DataObject.h"
#include "data/Parser.h"

string DataObject::serialize(const Parser &parser) const {
    return parser.serialize(this);
}

string DataObject::toString() const {
    string res = "{";

    for (const auto &it: children) {
        DataElement *element = it.second;
        res += '"' + it.first + "\":";
        res += element->toString();

        res += ",";
    }

    if (res.length() > 1)
        res.pop_back();

    res += "}";
    return res;
}